### ISSUE TITLE
Avoid clicking buttons to login with a single OAuth/OIDC access flow

### ISSUE THREAD

**Issue by smolinari** [coder/coder#6015]
Hi,

I just came across [this change](https://github.com/coder/coder/pull/5101) in the recent release and it is quite close to [the suggestion I made just the other day on Discord](https://discord.com/channels/747933592273027093/1070277782669639720) . 

However, it still isn't quite SSO. :grin: 

So, let me detail out the user story.....

As a user of an external IAM system and Coder, if I'm already logged in to the IAM system, I'd like to be able to just hit a URL of  Coder and gain access. I shouldn't need to click on a button, if the admin of both systems has only the one OIDC access flow set up between them. (even if there were two, the direct access should be possible, but that doesn't fit the reasoning of this request)

If I'm not logged in and happen to go to a Coder URL, Coder will redirect me to the (single) IAM system to log in, also without needing to click a button. The original URL should be "remembered" to get me back to the place I originally wanted to go in Coder (not sure of the feasibility here, as it depends on the IAM system).   

Basically, with this additional change, the idea is that Coder will be a smaller part of a larger system and users of that larger system can access Coder without needing to click an extra button to gain access, sort of like in an Intranet of a company. They login to start work and then any link to Coder will....just work. 

Scott

**Comment by BrunoQuaresma** [comment]
So just to check if I understand, if the user access coder and it is not logged in, and has GoogleOauth setup only, we should redirect to the Google url instead of showing the buttons. If the deployment has two providers setup, display the buttons. Is that right?

**Comment by smolinari** [comment]
That is correct. We have Keycloak as our IAM. Our internal devs/ devops, who might need to access the Coder UI to create "internal" workspaces, would be using Coder in this "Intranet-like" way. 

Scott

**Comment by bpmct** [comment]
We just added a feature to entirely disable user/password auth, so now there can only be one button, but I'm pretty hesitant for us to make `coder.example.com` directly redirect to your identity provider (you can customize the button logo and text though, see `coder server --help`).

If you can find some examples of other dev tool/enterprise software that supports this, perhaps we can reconsider, but it feels like an antipattern to me.

**Comment by smolinari** [comment]
Hi @bpmct Ben,

Thanks for the reply! To answer your question.....

[Saleforce.com](https://www.salesforce.com/) does it, when you set it up for SSO. It's because they know their customers are enterprises and that the work being done in Salesforce by the enterprise's team members is only a part of that organizations larger set of software and services used. 

[Jira](https://www.atlassian.com/software/jira) and [ServiceNow](https://www.servicenow.com/) also does this, when set up for SSO. I believe [Workday.com](https://www.workday.com/) would too, but while logging in, it shows an option for "non-sso users", which would be similar to Coder having both the login and the OIDC button showing. So, either we haven't turned off the non-sso option or Workday is thinking similar to you too. 

I know these to be facts, because I work with these services during my "normal" job every day. :grin: When I log in as a team member to my work computer, I get direct access to Salesforce, Jira and ServiceNow. No need to press any extra buttons. I just hit any page I want, and I'm "in". 

This use case covers the situation where the IAM/ OIDC provider is a single provider for an organization with multiple other services in parallel to Coder. 

It seems though you are right, that other OSS offerings also don't have this "point of view" about what the IAM/ OIDC provider is used for. 

For instance, we also have [Gitea ](https://gitea.io/en-us/)set up, and they too have the extra button to login. I made the same suggestion to their community, but nobody answered. I'll probably put in a suggestion to their Github repo too. I just haven't gotten to it, because it seem it will be probably futile, from what I gather of their progress there. They have much bigger issues to work on. 

We wanted to turn OIDC on for [Rancher](https://www.rancher.com/), but turning the OIDC connectivity on [broke the Rancher installation](https://github.com/rancher/rancher/pull/39310). So, I had to turn it back off and thus, I don't know how Rancher would offer its SSO, but for sure it should also just redirect. 

We are also working with [Mailu ](https://mailu.io/1.9/) about having OIDC added. I haven't gotten to asking for an automatic redirect from them. They are also a bit reluctant to introduce even OIDC, at least not how [it's been PRed](https://github.com/Mailu/Mailu/pull/2574), which is disappointing. 

But, in the future, imagine within our organization Gitea, Rancher, Mailu and Coder (and other services) would all be "internal" to our company, where only our devOps administrators would be needing access to these services to work on them. That is what we are intending to do or rather have. 

With all that said though, it's not really going to be deal breaker, if you don't want to make this feature a reality. The extra click of the button will just be a daily small annoyance to our devOps team members, because the login process isn't "pure" SSO. :grin: :shrug:      

Scott

**Comment by smolinari** [comment]
Quick update. Rancher finally got to fix their OIDC issue and it's the same as what Coder does. 

Mailu also came out with Version 2 and they do something completely different, where the proxy has to stop and check the user for auth. We haven't gotten into that one yet, but if it did work, that might also be a way to allow Coder to work too? Maybe?. I believe Mailu asks for the email of the user to be in the request header from the proxy (being its an email server). Hmmm, what could Coder require to open its doors. :thinking: 

Scott
